### DISCUSSION TITLE
TLS context - unclear when used?

### DISCUSSION THREAD

**Discussion by asfg84** [aws/aws-sdk-cpp#3005] [Q&A] [answered]
Dear all,

when trying to figure out why the initialization of the API via `initAPI(options)` takes so long, I found that the call to create the `Aws::Crt::Io::TlsContext` context is rather slow (https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Aws.cpp#L101). When digging deeper into the code, I could not find when this TLS context will ever be used again, because it is just used to create the TLS options which are then set via `SetDefaultTlsConnectionOptions` (https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Aws.cpp#L103 and https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Globals.cpp#L40) and live as a static member variable `g_defaultTlsConnectionOptions` in Globals.cpp. The only way this variable is accessed is via `GetDefaultTlsConnectionOptions()` (https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Globals.cpp#L43), which is never called from inside the aws cpp sdk. So I am wondering what this is about? Is there some usage of this static member that I cannot see? To me it does not seem to make a difference when I just comment the whole TLS initialization section, except that the call to initAPI() gets much faster.

Thanks!

**Comment by jmklix** [comment]
TLS context is only used by the s3 crt client. That being said it shouldn't be taking that long to initialize it. Can you provide a quick repo and details about your environment that can help us look into this? How long exactly is this taking to initialize with and without TLS context?

**Reply by asfg84** [reply, marked as answer]
Sorry for the late reply.  With this minimal example
```
  auto start = std::chrono::steady_clock::now();

  Aws::SDKOptions options;
  options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Info;
  Aws::InitAPI(options);

  auto dc = std::chrono::steady_clock::now() - start;
  auto ms = std::chrono::duration_cast<std::chrono::microseconds>(dc).count();

  printf("initialization runtime: %.6f ms", ms * 1e-3);
```
On an amd64 system with Ubuntu 22.04 I measure nearly 98ms initialization runtime with the TLS context, and around 32ms without it. (Of these 32ms, 31ms come from the initialization of the mqtt libraray, that we do not require either, see also https://github.com/awslabs/aws-crt-cpp/discussions/630).

The runtimes strongly depend on the system and architecture, of course. On amd64 with Ubuntu 20.04 I measure only 4ms with TLS and 2ms without, so this is no issue at all. But on armhf or amd64 and on Ubuntu 22.04 it is rather slow.
